src/java/org/apache/cassandra/index/sai/disk/v1/SegmentMetadataBuilder.java

```diff
@@ -0,0 +1,396 @@
+            this.values = values;
+            this.builder = builder;
+        }
+
+        @Override
+        public int getDocCount()
+        {
+            return values.getDocCount();
+        }
+
+        @Override
+        public long size()
+        {
+            return values.size();
+        }
+
+        @Override
+        public void getValue(int i, BytesRef packedValue)
+        {
+            values.getValue(i, packedValue);
+        }
+
+        @Override
+        public byte getByteAt(int i, int k)
+        {
+            return values.getByteAt(i, k);
+        }
+
+        @Override
+        public int getDocID(int i)
+        {
+            return values.getDocID(i);
+        }
+
+        @Override
+        public void swap(int i, int j)
+        {
+            values.swap(i, j);
+        }
+
+        @Override
+        public byte[] getMinPackedValue()
+        {
+            return values.getMinPackedValue();
+        }
+
+        @Override
+        public byte[] getMaxPackedValue()
+        {
+            return values.getMaxPackedValue();
+        }
+
+        @Override
+        public int getNumDimensions()
+        {
+            return values.getNumDimensions();
+        }
+
+        @Override
+        public int getBytesPerDimension()
+        {
+            return values.getBytesPerDimension();
+        }
+
+        @Override
+        public int getNumIndexDimensions() throws IOException
+        {
+            return values.getNumIndexDimensions();
+        }
+
+        @Override
+        public PointTree getPointTree() throws IOException
+        {
+            return values.getPointTree();
+        }
+
+        @Override
+        public void intersect(IntersectVisitor visitor) throws IOException
+        {
+            values.intersect((docId, term) -> {
+                if (!Arrays.equals(term, lastTerm))
+                {
+                    if (lastTerm != null)
+                        builder.add(encodeIfNeeded(lastTerm), count);
+
+
+                    count = 0;
+                    lastTerm = Arrays.copyOf(term, term.length);
+                }
+                count++;
+                visitor.visit(docId, term);
+            });
+        }
+
+        @Override
+        public void close() throws IOException
+        {
+            if (lastTerm != null)
+            {
+                builder.add(encodeIfNeeded(lastTerm), count);
+            }
+        }
+
+        /**
+         * Converts the term values coming from the index into a proper byte comparable values.
+         * This is needed because values of some types are stored in the index using their
+         * raw serializations, without encoding; but some others use byte comparable serialization.
+         */
+        private ByteComparable encodeIfNeeded(byte[] term)
+        {
+            AbstractType<?> type = builder.validator;
+            if (type instanceof InetAddressType || type instanceof IntegerType || type instanceof DecimalType)
```

Comment:
I think it is worth documenting that the `term` you're receiving has already run through the `TypeUtil.encode` method. Inet gets mapped from ivp4 to ipv6 and decimal is truncated to the 24 most significant bytes. These should be fine in the statistics logic since they mappings only lose precision but maintain order and are still valid representations of their types.

However, integer type (big int) gets this encoding:

```
    /**
     * Encode a {@link BigInteger} into a fixed width 20 byte encoded value.
     *
     * The encoded value is byte comparable and prefix compressible.
     *
     * The format of the encoding is:
     *
     *  The first 4 bytes contain the length of the {@link BigInteger} byte array
     *  with the top bit flipped for positive values.
     *
     *  The remaining 16 bytes contain the 16 most significant bytes of the
     *  {@link BigInteger} byte array.
     *
     *  For {@link BigInteger} values whose underlying byte array is less than
     *  16 bytes, the encoded value is sign extended.
     */
    public static ByteBuffer encodeBigInteger(ByteBuffer value)
```

At a glance, that projection seems like it could vary and cause problems, but I haven't had a chance to formally verify that yet. I think the main issue would come in the accuracy of range queries since point queries might encode okay. I'll have to check on Monday though.

Reply:
This method has been removed. We use the same byte comparable representation as the indexes now so we don't need to decode.